File: DoublyLinkedList.py
#Represent a node of doubly linked list    
class Node:    
    def __init__(self,data):    
        self.data = data;    
        self.previous = None;    
        self.next = None;    
            
class DoublyLinkedList:    
    #Represent the head and tail of the doubly linked list    
    def __init__(self):    
        self.head = None;    
        self.tail = None;    
            
    #addNode() will add a node to the list    
    def addNode(self, data):    
        #Create a new node    
        newNode = Node(data);    
            
        #If list is empty    
        if(self.head == None):    
            #Both head and tail will point to newNode    
            self.head = self.tail = newNode;    
            #head's previous will point to None    
            self.head.previous = None;    
            #tail's next will point to None, as it is the last node of the list    
            self.tail.next = None;    
        else:    
            #newNode will be added after tail such that tail's next will point to newNode    
            self.tail.next = newNode;    
            #newNode's previous will point to tail    
            newNode.previous = self.tail;    
            #newNode will become new tail    
            self.tail = newNode;    
            #As it is last node, tail's next will point to None    
            self.tail.next = None;    
                
    #display() will print out the nodes of the list    
    def display(self):    
        #Node current will point to head    
        current = self.head;    
        if(self.head == None):    
            print("List is empty");    
            return;    
        print("Nodes of doubly linked list: ");    
        while(current != None):     
            #Prints each node by incrementing pointer.    
            print(current.data),;    
            current = current.next;    
                
dList = DoublyLinkedList();    
#Add nodes to the list    
dList.addNode(1);    
dList.addNode(2);    
dList.addNode(3);    
dList.addNode(4);    
dList.addNode(5);    
     
#Displays the nodes present in the list    
dList.display();

File: Doubly_Linked_List.cpp
#include <iostream> 
using namespace std; 

struct node 
{
    int data; 
    struct node *next; 
    struct node *prev;
    
}; 

struct node *head = NULL; 
struct node *temp; 
struct node *tail;
struct node *current; 
struct node *nextnode; 
int count = 0;
int pos; 
int length; 



//IMPLEMENTING

void implement(){

    int choice; 
 
    while(choice){

        node *newnode = new node; 
        cout << "enter the data u want to add: " << endl; 
        cin >> newnode -> data; 
        newnode -> next = 0; 
        newnode -> prev = 0; 

        if(head == NULL){
            head = temp = tail = newnode; 

        }else {
            temp -> next = newnode; 
            newnode -> prev = temp; 
            temp = newnode; 
            tail = temp; 
        }

        cout << "do u want to continue" <<endl; 
        cin >> choice; 

    }
   
    
}



//INSERTION AT BEGINING 

void ins_at_beg(){
    node *newnode = new node; 
    cout << "enter the data u want to insert at begining:  " <<endl; 
    cin >> newnode -> data ;

    newnode -> next = head; 
    head -> prev = newnode; 
    head = newnode; 

}



//INSERTION AT END 

void ins_at_end(){
    node *newnode = new node; 
    cout << "enter the data u want to insert at end:  " <<endl; 
    cin >> newnode -> data ;
    if(head == 0){
        head = tail = newnode; 

    }
    else {
   
        tail -> next = newnode; 
        newnode -> prev = tail; 
        tail = newnode; 

    }
   
}



//INSERTION AT A POSITION

void ins_at_pos(){
    

    node *newnode = new node; 
    cout << "enter the position u want to insert in: " <<endl; 
    cin >> pos; 

    cout << "enter the data u want to insert at that position: " <<endl; 
    cin >> newnode -> data; 

    if(pos>length){
        cout << "invalid posiiton" <<endl; 
    }
    else if(pos==1){
        ins_at_beg(); 
    }
    
    else {
        int i = 1; 
        temp = head; 
         while(i < pos-1){
        temp = temp -> next; 
        i++; 

        }
        newnode -> prev = temp; 
        newnode -> next = temp -> next; 
        temp -> next = newnode; 
        newnode -> next -> prev = newnode;  

    }

}



//INSERTION AFTER A POSITION 

void ins_after_pos(){
    

    node *newnode = new node; 
    cout << "enter the position after which u want to insert in: " <<endl; 
    cin >> pos; 

    cout << "enter the data u want to insert at that position: " <<endl; 
    cin >> newnode -> data; 

    if(pos>length){
        cout << "invalid posiiton" <<endl; 
    }
    else if(pos == 1){
        ins_at_beg(); 
    }
    
    else {
        int i = 1; 
        temp = head; 
         while(i < pos){
        temp = temp -> next; 
        i++; 

        }
        newnode -> prev = temp; 
        newnode -> next = temp -> next; 
        temp -> next = newnode; 
        newnode -> next -> prev = newnode;  

    }


}


//DELETION FROM BEGINING 

void del_beg(){
    if(head == NULL){
        cout << "list is empty: " <<endl; 

    }
    else {

        head = head -> next; 
        head -> prev = NULL; 
        delete temp; 
    }
}



//DELETION FROM END

void del_end(){
    if(head == NULL){
        cout << "list is empty" <<endl; 
    }
    else {
        temp = tail; 
        tail = tail->prev; 
        tail -> next = NULL; 
        delete temp; 
    }
}


//DELETION FROM A POSITION

void del_pos(){

    cout << "enter the position that you want to delete from: " <<endl; 
    cin >> pos; 

    if(head==NULL){
        cout << "list empty: " <<endl; 
    }
    else if(pos > length){
        cout << "invalid position" <<endl;
    }
    {   temp = head; 
        int i = 1; 
        while(i < pos){
            temp = temp->next; 
            i++;
        }

        temp -> prev -> next = temp -> next; 
        temp -> next -> prev = temp -> prev; 
        delete temp; 
    }
}


//REVERSE 

void reverse(){
    if(head == NULL){
        cout << "empty list" <<endl; 
    }
    else{
        current = head;
        while(current!=0){
            nextnode =  current -> next; 
            current -> next = current -> prev; 
            current -> prev = nextnode; 
            current = nextnode; 
        }

        current = head; 
        head = tail; 
        tail = current; 

    }


}


void display(struct node *temp, struct node *head){
    cout << "here is your linked list: " << endl; 
    int count = 0; 
    temp = head; 
    while(temp!=NULL){

        cout << temp -> data << " " ;
        temp = temp-> next;
        count++ ; 
        length = count; 

    }
    cout << endl; 

    cout << "tot. no. of elements in the list now: " << count << endl; 

    cout << endl; 
    
}

//SAMPLE CASES

int main(){

    implement();
    display(temp, head);

    ins_at_beg();
    display(temp, head);

    ins_at_end();
    display(temp, head);

    ins_at_pos();
    display(temp, head);

    ins_after_pos();
    display(temp, head);

    del_beg();
    display(temp, head);

    del_end();
    display(temp, head);
    
    del_pos();
    display(temp, head);

    reverse();
     display(temp, head);



}
